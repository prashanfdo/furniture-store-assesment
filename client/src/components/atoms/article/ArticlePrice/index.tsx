import { DEFAULT_LOCALE_NUMBER_FORMAT } from "helpers/constants";
import { useMemo } from "react";

type ArticlePriceProps = {
  currency: string;
  price: number;
  salePrice?: number;
  discount?: number;
};
const ArticlePrice: React.FC<ArticlePriceProps> = ({ currency, price, salePrice, discount }: ArticlePriceProps) => {
  const [finalPrice, regularPrice, discountPercent] = useMemo(() => {
    const curFormatter = new Intl.NumberFormat(DEFAULT_LOCALE_NUMBER_FORMAT, { style: "currency", currency });
    const finalPrice = curFormatter.format(salePrice || price);
    const regularPrice = curFormatter.format(price);
    const discountPercent = Math.round(((discount || 0) / price) * 100);
    return [finalPrice, regularPrice, discountPercent];
  }, [salePrice, price, discount, currency]);

  return (
    <div className="text-gray-600">
      <span data-testid="final-price" className="font-semibold text-red-600">
        {finalPrice}
      </span>
      {discountPercent > 0 ? (
        <span data-testid="regular-price" className="ml-4 text-sm line-through">
          {regularPrice}
        </span>
      ) : null}
      {discountPercent > 0 ? (
        <span data-testid="discount-percent" className="px-1 py-1 ml-2 text-sm text-white bg-red-500 rounded-lg">
          -{discountPercent}%
        </span>
      ) : null}
    </div>
  );
};

export default ArticlePrice;
