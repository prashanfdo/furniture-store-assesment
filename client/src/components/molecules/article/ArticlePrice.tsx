import { useMemo } from "react";

type ArticlePriceProps = {
  currency: string;
  price: number;
  salePrice?: number;
  discount?: number;
};
const ArticlePrice: React.FC<ArticlePriceProps> = ({
  currency,
  price,
  salePrice,
  discount,
}: ArticlePriceProps) => {
  const [finalPrice, regularPrice, discountPercent] = useMemo(() => {
    const curFormatter = new Intl.NumberFormat("de-DE", { style: "currency", currency });
    const finalPrice = curFormatter.format(salePrice || price);
    const regularPrice = curFormatter.format(price);
    const discountPercent = Math.round(((discount || 0) / price) * 100);
    return [finalPrice, regularPrice, discountPercent];
  }, [salePrice, price, discount, currency]);
  return (
    <div className="text-gray-600">
      <span className="text-red-600 font-semibold">{finalPrice}</span>
      {discountPercent > 0 ? (
        <span className="line-through text-sm ml-4">{regularPrice}</span>
      ) : null}
      {discountPercent > 0 ? (
        <span className="bg-red-500 text-sm ml-2 rounded-lg py-1 px-1 text-white">
          -{discountPercent}%
        </span>
      ) : null}
    </div>
  );
};

export default ArticlePrice;
