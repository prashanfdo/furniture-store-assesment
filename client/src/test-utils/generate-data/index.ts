export * from "./article";
export * from "./category";
export * from "./search";
