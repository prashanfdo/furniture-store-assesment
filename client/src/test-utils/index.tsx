import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
