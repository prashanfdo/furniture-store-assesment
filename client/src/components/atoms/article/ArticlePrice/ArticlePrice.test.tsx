import { render, screen } from "test-utils";
import ArticlePrice from "./index";

describe("ArticlePrice", () => {
  it("should render", () => {
    const { container } = render(<ArticlePrice currency="EUR" price={100} discount={10} salePrice={90} />);
    expect(container).toMatchSnapshot();
  });
  it("should render with discount", () => {
    render(<ArticlePrice currency="EUR" price={100} discount={10} salePrice={90} />);
    expect(screen.getByTestId("final-price")).toHaveTextContent("90,00 €");
    expect(screen.getByTestId("regular-price")).toHaveTextContent("100,00 €");
    expect(screen.getByTestId("discount-percent")).toHaveTextContent("-10%");
    expect(getComputedStyle(screen.getByTestId("regular-price")).textDecorationLine).toContain("line-through");
  });
  it("should render with no discount", () => {
    render(<ArticlePrice currency="EUR" price={100} />);
    expect(screen.getByTestId("final-price")).toHaveTextContent("100,00 €");
    expect(screen.queryByTestId("regular-price")).not.toBeInTheDocument();
    expect(screen.queryByTestId("discount-percent")).not.toBeInTheDocument();
  });
});
