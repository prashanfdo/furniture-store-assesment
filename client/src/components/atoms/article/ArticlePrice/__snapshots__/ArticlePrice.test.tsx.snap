// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlePrice should render 1`] = `
<div>
  <div
    class="text-gray-600"
  >
    <span
      class="font-semibold text-red-600"
      data-testid="final-price"
    >
      90,00 €
    </span>
    <span
      class="ml-4 text-sm line-through"
      data-testid="regular-price"
    >
      100,00 €
    </span>
    <span
      class="px-1 py-1 ml-2 text-sm text-white bg-red-500 rounded-lg"
      data-testid="discount-percent"
    >
      -
      10
      %
    </span>
  </div>
</div>
`;
